avr-hal-generic/src/i2c.rs

```diff
@@ -599,73 +593,137 @@ macro_rules! impl_twi_i2c {
                 sda: $sdamod::$SDA<$crate::port::mode::Input<$crate::port::mode::Floating>>,
                 scl: $sclmod::$SCL<$crate::port::mode::Input<$crate::port::mode::Floating>>,
                 address: u8,
-            ) -> [<$I2c Slave>]<CLOCK, $crate::i2c::I2cFloating> {
+            ) -> [<$I2c Slave>]<$crate::i2c::I2cFloating> {
                 [<$I2c Slave>] {
                     p,
                     sda,
                     scl,
-                    _clock: ::core::marker::PhantomData,
                     address,
                 }
             }
         }
 
-        impl<CLOCK, M> [<$I2c Slave>]<CLOCK, M>
-        where
-            CLOCK: $crate::clock::Clock,
-        {
-            /// Initialize the slave on the I2C bus.
-            ///
-            /// # Arguments
-            /// 
-            /// * `gce` - if `true`, then listen on the General Call Address.
-            ///
-            /// # Note
-            ///
-            /// The general use case is to pass `false` here and only listen
-            /// for messages addressed to this slave address.
-            pub fn start(&mut self, gce: bool) {
-                let gce_mask = if gce {1} else {0};
-                let rawaddr = (self.address << 1) | gce_mask; 
-                self.p.$twar.write(|w| unsafe {w.bits(self.address)});
-                self.p.$twcr.write(|w| w
-                    .$twen().set_bit()
-                    .$twea().set_bit()
-                    .$twsta().clear_bit()
-                    .$twsto().clear_bit()
-                    .$twint().set_bit()
-                );
+
+        // Slave State Machine
```

Comment:
I wrote a simplified version of what `lpc8xx-hal` has. I may eventually end up with exactly what they wrote, but I'm just trying to build up from scratch with the simplest API possible. I researched other popular rust state machine patterns, and settled on this.

I like the idea of using the compiler and type checker to validate the state transitions, but this looks ugly to me:

https://hoverbear.org/blog/rust-state-machine-pattern/

I like the all transitions listed in one place as here:

https://dev.to/werner/practical-rust-web-development-state-machine-2jlc

But I prefer to have use types with explicit methods on them as here:

     TypeA.fooMethod() -> TypeB
     TypeB.barMethod() -> TypeC

etc.

I'm just stubbing the methods out here to get your thoughts before I flesh out the implementation.